File: aptoide-database/src/main/java/cm/aptoide/pt/database/room/DownloadDAO.java
package cm.aptoide.pt.database.room;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import io.reactivex.Observable;
import io.reactivex.Single;
import java.util.List;

import static androidx.room.OnConflictStrategy.REPLACE;

@Dao public interface DownloadDAO {

  @Query("SELECT * from download") Observable<List<RoomDownload>> getAll();

  @Query("SELECT * from download where md5 = :md5 LIMIT 1 ") Single<RoomDownload> getAsSingle(
      String md5);

  @Query("SELECT * from download where md5 = :md5 LIMIT 1 ")
  Observable<RoomDownload> getAsObservable(String md5);

  @Query("DELETE from download where md5= :md5") void remove(String md5);

  @Query("DELETE from download where packageName=:packageName and versionCode=:versionCode")
  void remove(String packageName, int versionCode);

  @Insert(onConflict = REPLACE) void insertAll(List<RoomDownload> downloads);

  @Insert(onConflict = REPLACE) void insert(RoomDownload download);

  @Query("SELECT * from download where overallDownloadStatus = "
      + RoomDownload.PROGRESS
      + " OR overallDownloadStatus = "
      + RoomDownload.IN_QUEUE
      + " OR overallDownloadStatus = "
      + RoomDownload.PENDING) Observable<List<RoomDownload>> getRunningDownloads();

  @Query("SELECT * from download where overallDownloadStatus="
      + RoomDownload.IN_QUEUE
      + " ORDER BY timeStamp ASC") Observable<List<RoomDownload>> getInQueueSortedDownloads();

  @Query("SELECT * from download where md5 = :md5") Observable<List<RoomDownload>> getAsList(
      String md5);

  @Query("SELECT * from download where overallDownloadStatus="
      + RoomDownload.ERROR
      + " and downloadError="
      + RoomDownload.NOT_ENOUGH_SPACE_ERROR
      + " ORDER BY timeStamp ASC") Observable<List<RoomDownload>> getOutOfSpaceDownloads();
}

File: downloads_database/src/main/java/cm/aptoide/pt/downloads_database/data/database/DownloadDao.kt
package cm.aptoide.pt.downloads_database.data.database

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import cm.aptoide.pt.downloads_database.data.database.model.DownloadEntity
import io.reactivex.Observable
import io.reactivex.Single

@Dao
interface DownloadDao {

  @Query("SELECT * from download")
  fun getAllDownloads(): Observable<List<DownloadEntity>>

  @Query("SELECT * from download where md5 = :md5 LIMIT 1 ")
  fun getDownload(md5: String): Single<DownloadEntity>

  @Query("SELECT * from download where md5 = :md5 LIMIT 1 ")
  fun observeDownload(md5: String): Observable<DownloadEntity>

  @Query("DELETE from download where md5= :md5")
  fun removeDownload(md5: String)

  @Insert(onConflict = OnConflictStrategy.REPLACE)
  fun saveDownload(downloadEntity: DownloadEntity)

  @Query(
    "SELECT * from download where overallDownloadStatus = "
        + DownloadEntity.PROGRESS
        + " OR overallDownloadStatus = "
        + DownloadEntity.IN_QUEUE
        + " OR overallDownloadStatus = "
        + DownloadEntity.PENDING
  )
  fun getRunningDownloads(): Observable<List<DownloadEntity>>

  @Query(
    "SELECT * from download where overallDownloadStatus="
        + DownloadEntity.IN_QUEUE
        + " ORDER BY timeStamp ASC"
  )
  fun getInQueueDownloads(): Observable<List<DownloadEntity>>

  @Query(
    "SELECT * from download where overallDownloadStatus="
        + DownloadEntity.WAITING_TO_MOVE_FILES
        + " ORDER BY timeStamp ASC"
  )
  fun getUnmovedDownloads(): Observable<List<DownloadEntity>>

  @Query("SELECT * from download where packageName = :packageName AND overallDownloadStatus=" + DownloadEntity.COMPLETED + " LIMIT 1 ")
  fun getCompletedDownload(packageName: String): Observable<DownloadEntity>

}